[AOTInductor] ABI-Compatibility for RecordFunction.

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160380
* #160131
* __->__ #159842

Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

muchulee8 committed:
[AOTInductor] ABI-Compatibility for RecordFunction.

Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D79622071).

iremyux commented:
Adding `ciflow/win-arm64` label to trigger Windows Arm64 CI.
(This is for testing purposes only and isn’t related to any changes in this PR.)

muchulee8 committed:
Update on "[AOTInductor] ABI-Compatibility for RecordFunction."


Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] ABI-Compatibility for RecordFunction."


Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] ABI-Compatibility for RecordFunction."


Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] ABI-Compatibility for RecordFunction."


Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -5491,7 +5491,10 @@ def codegen_group(self, name=None) -> str:
-                        f'RECORD_FUNCTION("{prefix + kernel_name}", c10::ArrayRef<c10::IValue>({{}}));'
+                        (
+                            "torch::aot_inductor::RAIIAtenRecordFunctionHandle "
+                            f'record_{prefix + kernel_name}_("{prefix + kernel_name}", nullptr);'
```
Is it true that right now kwargs are is always empty?

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1056,6 +1057,38 @@ AOTITorchError aoti_torch_check_inf_and_nan(
+      auto wrappedKwargs =
+          reinterpret_cast<std::unordered_map<std::string, C10IValueHandle>*>(
+              kwargs);
+      for (const auto& pair : *wrappedKwargs) {
```
Back to earlier comment about kwargs, if kwargs is not null, then we need a shim function to construct such a map, right?

muchulee8 replied:
Yes that's correct. We won't be using c10:: namespace directly, and would be fully relying on similar mechanism we had for AtenTensorHandle.

muchulee8 committed:
Update on "[AOTInductor] ABI-Compatibility for RecordFunction."


Summary:
Previous our implementation for RecordFunction injects Aten into
codegen, which is breaking the ABI contract for AOTInductor.

C10::IValue is aded to call the full record function. The extension of
more profiling info will come in later PRs.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov coconutruben

Differential Revision: [D79622071](https://our.internmc.facebook.com/intern/diff/D79622071)

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot rebase

muchulee8 committed:
Update

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot merge -f "merged internally, win arm64 failure unrelated"
